Modules/IO/ImageBase/include/itkImageFileWriter.h

```diff
@@ -235,6 +235,32 @@ class ITK_TEMPLATE_EXPORT ImageFileWriter : public ProcessObject
   int  m_CompressionLevel{ -1 };
   bool m_UseInputMetaDataDictionary{ true };
 };
+
+template <>
+class ImageFileWriter<void>
+{
+public:
+  ITK_DISALLOW_COPY_AND_MOVE(ImageFileWriter);
+
+  ImageFileWriter() = delete;
+  ~ImageFileWriter() = delete;
+
+  /** Writes an image to the specified file. Example:
+    \code
+    itk::ImageFileWriter<>::WriteImage(*imagePointer, outputFileName);
+    \endcode
+   */
+  template <typename TInputImage>
+  static void
+  WriteImage(const TInputImage & image, const std::string & fileName)
```

Comment:
?? Should we expose other conditional behaviors as default options?

 UseCompression = false, CompressionLevel = -1, UseINputMetaDataDictionary=true)

Reply:
Good question. The aim of this PR is to make the most common use case a simple "one-liner".

Reply:
My solution does not change the most common use case, but does allow for more flexibility easily for those who need it.  The UseCompression is the issue I am most interested in maintaining access to.

Reply:
`itk::ImageFileWriter<ImageType>` is being used by ITK itself more than 500 times. I wonder how many times (%) they just assume the default values for those options.

Reply:
I prefer to have two overloads, one for raw pointer one for smart pointer. That way we can just provide image variable name at the point of invocation, see https://github.com/dzenanz/ITKMorphologicalContourInterpolation/commit/a9ac128f57078cb7c71d1abacd1ce8fce02f1c96#diff-2c8d1766ca920beada092bae39ef4684594b715a9eb835bcc35ad15998b2c86aR1339.

Note that `itk::Image<bool, dim>` needs a distinct overload with my approach. As that is not commonly used, I am not sure whether we want to support it here.

Reply:
I agree with Hans. I frequently need to write both compressed and uncompressed images. When writing label maps, I always enable compression. With intensity images I mostly disable it, as it is much faster and takes only about 2x as much as compressed ones.

Reply:
Of course, extra options could still be added as function parameters with default argument values. (Possibly afterwards.)

Another redesign could be to add those options as data members to the new `itk::ImageFileWriter<void>`, very much like the original `ImageFileWriter<ImageType>`, but then still support a "one-liner" like:

    itk::ImageFileWriter<>{}.WriteImage(image, fileName);

Which would default-construct a writer (containing default values for its options), and then use it directly, calling a non-static `WriteImage` member function that takes the options into account. Or classic ITK `New()` style:

    itk::ImageFileWriter<>::New()->WriteImage(image, fileName);

For the moment I still prefer the current PR (just a static member function), possible with some optional default arguments. What do you think?

Reply:
If the proposed option is implemented `UseCompression = false, CompressionLevel = -1, UseINputMetaDataDictionary=true` then we will have calls like:
```
itk::ImageFileWriter<void>::WriteImage(image, "something.ext", true, 9, true);
```
In C++ the unnamed parameters are not clear what the argument position means.

Perhaps adding a `WriteImageWithCompression` function would satisfy this added requirements?

> I prefer to have two overloads, one for raw pointer one for smart pointer.

There should be implicit conversions from smart pointers to raw pointers, therefor just a method with a raw pointer should work.

This is a new style for ITK, we have not had function like calls before for process object. An alternative for the void template parameters would be to place the template function(s) into a namespace.

Reply:
> There should be implicit conversions from smart pointers to raw pointers

Maybe, but if there is no exact match template arguments cannot be deduced.

> place the template function(s) into a namespace

I like it! So we could have:
```cpp
#include "ImageFileWriter.h"
...
itk::WriteImage(image, fileName);
itk::WriteImage(filter->GetOutput(), fileName2, true); // compressed
```
I like default `UseCompression = false` better than `WriteImageWithCompression`, but I could live with the latter too.

Reply:
I like the idea of having a `WriteImage` in `itk` namespace, rather than as a static member of  `itk::ImageFileWriter<void>`. The only (minor) drawback could be that it might be less obvious where to define `itk::WriteImage`. Should it then be in its own "itkWriteImage.h"?

Reply:
It would have to include `itkImageFileWriter.h` anyway, so a separate include file would reduce its convenience somewhat. @blowekamp what is your preference?